### PR TITLE
Issue #7689: update doc for BooleanExpressionComplexity

### LINKED ISSUE #7689: update doc for BooleanExpressionComplexity
https://checkstyle.org/config_metrics.html#BooleanExpressionComplexity

Add example for each existing config to show what code is ok and what code cause violation.
Good example is https://checkstyle.org/config_naming.html#AbstractClassName
Please add examples for all properties of Check.

look at referenced commits of #5832, to see what and how should be changed in code.

please play with module/Check by means of https://checkstyle.org/cmdline.html to understand how it works.

------------

PLEASE READ https://github.com/checkstyle/checkstyle/wiki/Good-practice-on-how-to-update-examples

### PR DESCRIPTION
**Pull request by akki111singh**
Fixes #7689

### CODE REVIEW

**Comment by akki111singh** [comment]
Current look
Screenshot from 2020-03-01 16-02-58
Screenshot from 2020-03-01 16-03-12
Screenshot from 2020-03-01 16-03-21

**Comment by akki111singh** [comment]
Output of the deafult example:

$ cat config.xml

<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
      "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
      "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="BooleanExpressionComplexity"/>
  </module>
</module>

$ cat Test.java

public class Test
{
    public static void main(String args[])
    {
        boolean a = true;
        boolean b = false;

        boolean c = (a & b) | (b & a); //OK
        boolean d = (!a && b) ^ (a || !b) && a; //violation
    }
}
$ java -jar checkstyle-8.30-all.jar -c config.xml Test.java

Starting audit...
[ERROR] /home/akhil/GSOC/Test.java:9:19: Boolean expression complexity is 4 (max allowed is 3). [BooleanExpressionComplexity]
Audit done.
Checkstyle ends with 1 errors.

**Comment by akki111singh** [comment]
Output of Non default examples:

CONFIG 1:

$ cat config1.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
      "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
      "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="BooleanExpressionComplexity">
      <property name="max" value="7"/>
    </module>
  </module>
</module>

$ cat Test1.java
public class Test
{
    public static void main(String args[])
    {
        boolean a = true;
        boolean b = false;

        boolean c = (a & b) | (b & a) ^ (a & b) | (b & a); //OK
        boolean d = (!a | b) ^ (a | !b) ^ (a || b) | (b & a) ^ b; //violation
    }
}
$ java -jar checkstyle-8.30-all.jar -c config1.xml Test1.java
Starting audit...
[ERROR] /home/akhil/GSOC/Test1.java:9:19: Boolean expression complexity is 8 (max allowed is 7). [BooleanExpressionComplexity]
Audit done.
Checkstyle ends with 1 errors.


Config2:

cat config2.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
     "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
     "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
 <module name="TreeWalker">
   <module name="BooleanExpressionComplexity">
     <property name="tokens" value="BXOR,LAND,LOR"/>
   </module>
 </module>
</module>

$ cat Test2.java
public class Test
{
   public static void main(String args[])
   {
       boolean a = true;
       boolean b = false;

       boolean d = (!a && b) | (a || !b) && a; // OK as | is ignored
       boolean c = (a && b) || (b && a) && a; //violation
   }
}
$ java -jar checkstyle-8.30-all.jar -c config2.xml Test2.java
Starting audit...
[ERROR] /home/akhil/GSOC/Test2.java:9:19: Boolean expression complexity is 4 (max allowed is 3). [BooleanExpressionComplexity]
Audit done.
Checkstyle ends with 1 errors.

**Comment by akki111singh** [comment]
@romani Can you please review it. Thanks!

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java [hunk 1]
```diff
@@ -74,6 +74,22 @@
...
+ * <pre>
+ * public class Test
+ * {
+ *  public static void main(String args[])
+ *  {
+ *    boolean a = true;
+ *    boolean b = false;
+ *
+ *    boolean c = (a &amp; b) | (b &amp; a); //OK
+ *    boolean d = (!a &amp;&amp; b) ^ (a || !b) &amp;&amp; a; //violation
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java]: Please change the comment to `// violation: complexity is 4` for clarity. 
Here and below.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java [hunk 2]
```diff
@@ -74,6 +74,22 @@
  * <pre>
  * &lt;module name="BooleanExpressionComplexity"/&gt;
  * </pre>
+ * <p>Code Example:</p>
+ * <div class="wrapper">
+ * <pre>
+ * public class Test
+ * {
+ *  public static void main(String args[])
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java]: Please change `String args[]` -> `String... args` to make it less c-stylish.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java (hunk 1 above)
Comment (akki111singh) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java (hunk 2 above)
Comment (akki111singh) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java [hunk 3]
```diff
@@ -74,6 +74,22 @@
  * <pre>
  * &lt;module name="BooleanExpressionComplexity"/&gt;
  * </pre>
+ * <p>Code Example:</p>
+ * <div class="wrapper">
+ * <pre>
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java]: Javadocs should only have `pre` tag, not `div`.

File: src/xdocs/config_metrics.xml [hunk 4]
```diff
@@ -101,6 +101,22 @@
         <source>
 &lt;module name=&quot;BooleanExpressionComplexity&quot;/&gt;
         </source>
+        <p>Code Example:</p>
+        <div class="wrapper">
+          <pre>
```
Comment (rnveach) [inline src/xdocs/config_metrics.xml]: xdocs should have `source` tag not, `div` or `pre`.

File: src/xdocs/config_metrics.xml [hunk 5]
```diff
@@ -101,6 +101,22 @@
...
+            public class Test
```
Comment (rnveach) [inline src/xdocs/config_metrics.xml]: All examples should start at the first column and not start off this far indented. Look at the screenshot and how the example is so far to the right.

File: src/xdocs/config_metrics.xml [hunk 6]
```diff
@@ -111,6 +127,22 @@
...
+            <pre>
+              public class Test
+              {
+                public static void main(String ... args)
+                {
+                  boolean a = true;
+                  boolean b = false;
+
+                  boolean c = (a &amp; b) | (b &amp; a) ^ (a &amp; b) | (b &amp; a); //OK
+                  boolean d = (a|b)^(a|b)^(a || b)|(b &amp; a)^ b; //violation,complexity is 8
```
Comment (rnveach) [inline src/xdocs/config_metrics.xml]: Please space out this line so it is easier to read. Everything is so compact. Line is allowed to span multiple if needed.

File: src/xdocs/config_metrics.xml [hunk 7]
```diff
@@ -101,6 +101,22 @@
...
+          <pre>
+            public class Test
+            {
+              public static void main(String ... args)
+              {
+                boolean a = true;
+                boolean b = false;
+
+                boolean c = (a &amp; b) | (b &amp; a); //OK
+                boolean d = (!a &amp;&amp; b) ^ (a || !b) &amp;&amp; a; //violation,complexity is 4
```
Comment (rnveach) [inline src/xdocs/config_metrics.xml]: Please explain on a subsequent line where the counts come from. It helps to explain to users how counting happens so they grasp the check easier.

Example: `X, Y, and Z are counted as 1 each which gives the total of 3. A, B, and C are ignored which don't add to the score`.
I would even consider adding counting to non-violations to show it is below the threshold.

File: src/xdocs/config_metrics.xml (hunk 4 above)
Comment (akki111singh) [reply src/xdocs/config_metrics.xml]: Screenshot from 2020-03-03 04-09-44

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java (hunk 3 above)
Comment (akki111singh) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java]: done

File: src/xdocs/config_metrics.xml (hunk 4 above)
Comment (akki111singh) [reply src/xdocs/config_metrics.xml]: done

File: src/xdocs/config_metrics.xml (hunk 5 above)
Comment (akki111singh) [reply src/xdocs/config_metrics.xml]: Current Look:
Screenshot_2020-03-03 checkstyle – Metrics(1)
Screenshot_2020-03-03 checkstyle – Metrics(2)
Screenshot_2020-03-03 checkstyle – Metrics(3)

File: src/xdocs/config_metrics.xml (hunk 6 above)
Comment (akki111singh) [reply src/xdocs/config_metrics.xml]: done

File: src/xdocs/config_metrics.xml (hunk 7 above)
Comment (akki111singh) [reply src/xdocs/config_metrics.xml]: done

File: src/xdocs/config_metrics.xml (hunk 5 above)
Comment (rnveach) [reply src/xdocs/config_metrics.xml]: @pietern please update your initial post with the current screenshot .

File: src/xdocs/config_metrics.xml [hunk 8]
```diff
@@ -101,6 +101,23 @@
...
+        <source>
+public class Test
+{
+  public static void main(String ... args)
+  {
+    boolean a = true;
+    boolean b = false;
+
+    boolean c = (a &amp; b) | (b &amp; a);
+    //OK, (&amp;,|) are counted as 1 which gives the total complexity of 3(max allowed 3)
```
Comment (rnveach) [inline src/xdocs/config_metrics.xml]: 1 space after all comments, it makes them easier to read.

> `(&amp;,|)`

I would not use parenthesis or commas as it could be confused with the actual parenthesis in the code. You can use the technical name for the operators or stick with `&` and `|` but again, space them apart so they aren't so compact. The comment can span multiple lines if needed.

> total complexity of 3(max allowed 3)

Again, space between words and `(`.

File: src/xdocs/config_metrics.xml [hunk 9]
```diff
@@ -101,6 +101,23 @@
...
+
+    boolean d = (!a &amp;&amp; b) ^ (a || !b) &amp;&amp; a;
+    //violation, (&amp;&amp;,^,||) are counted as 1 which gives total of 4(max allowed 3)
```
Comment (rnveach) [inline src/xdocs/config_metrics.xml]: same.

File: src/xdocs/config_metrics.xml [hunk 10]
```diff
@@ -111,17 +128,52 @@
...
+          <source>
+public class Test
+{
+  public static void main(String ... args)
+  {
+    boolean a = true;
+    boolean b = false;
+
+    boolean c = (a &amp; b) | (b &amp; a) ^ (a &amp; b) | (b &amp; a);
+    //OK, since the complexity here is 7(max allowed 7) as (&amp;,|,^) are counted as 1.
```
Comment (rnveach) [inline src/xdocs/config_metrics.xml]: same.

File: src/xdocs/config_metrics.xml [hunk 11]
```diff
@@ -111,17 +128,52 @@
...
+
+    boolean d = (a | b) ^ (a | b) ^ (a || b) | (b &amp; a) ^ b;
+    //violation,complexity here is 8 (max allowed 7) as (|,^,&amp;,||) are counted as 1.
```
Comment (rnveach) [inline src/xdocs/config_metrics.xml]: same.

File: src/xdocs/config_metrics.xml [hunk 12]
```diff
@@ -111,17 +128,52 @@
...
           <code>|</code>:
         </p>
+
         <source>
 &lt;module name="BooleanExpressionComplexity"&gt;
   &lt;property name="tokens" value="BXOR,LAND,LOR"/&gt;
 &lt;/module&gt;
-        </source>
-
+      </source>
```
Comment (rnveach) [inline src/xdocs/config_metrics.xml]: `</source>` tag looks to not be indented as far over as it was before create an unnecessary change. The other tags should align with this tag too.

File: src/xdocs/config_metrics.xml [hunk 13]
```diff
@@ -111,17 +128,52 @@
...
+        <source>
+  public class Test
+  {
+    public static void main(String ... args)
+    {
+      boolean a = true;
+      boolean b = false;
+
+      boolean c = (a || b) || (b &amp;&amp; a) || a;
+      //violation,complexity is 4 (max allowed 3) as (||,&amp;&amp;) are counted as 1;
```
Comment (rnveach) [inline src/xdocs/config_metrics.xml]: same.

File: src/xdocs/config_metrics.xml [hunk 14]
```diff
@@ -111,17 +128,52 @@
...
+
+      boolean d = (!a &amp;&amp; b) | (a || !b) &amp;&amp; a;
+      // OK,since complexity here is 3(max allowed 3) as (&amp;&amp;,||) are counted as 1,
+      // | is ignored here because tokens to check here are (&amp;&amp;,||,^)
```
Comment (rnveach) [inline src/xdocs/config_metrics.xml]: same.

File: src/xdocs/config_metrics.xml (hunk 9 above)
Comment (rnveach) [inline src/xdocs/config_metrics.xml]: The comment `// violation` should be on the same line as it would be reported in checkstyle. If more room is needed to explain reasoning, then that can be on the next line in its own comment.

Comment (akki111singh) [reply src/xdocs/config_metrics.xml]: @rnveach can i use multiline comments like /* */
Also if // violation and // ok should be on same line then the next line comment should begin from start of the line or below it

File: src/xdocs/config_metrics.xml (hunk 8 above)
Comment (akki111singh) [reply src/xdocs/config_metrics.xml]: Done

File: src/xdocs/config_metrics.xml (hunk 9 above)
Comment (akki111singh) [reply src/xdocs/config_metrics.xml]: Done

File: src/xdocs/config_metrics.xml (hunk 10 above)
Comment (akki111singh) [reply src/xdocs/config_metrics.xml]: Done

File: src/xdocs/config_metrics.xml (hunk 11 above)
Comment (akki111singh) [reply src/xdocs/config_metrics.xml]: Done

File: src/xdocs/config_metrics.xml (hunk 12 above)
Comment (akki111singh) [reply src/xdocs/config_metrics.xml]: Done

File: src/xdocs/config_metrics.xml (hunk 13 above)
Comment (akki111singh) [reply src/xdocs/config_metrics.xml]: Done

File: src/xdocs/config_metrics.xml (hunk 14 above)
Comment (akki111singh) [reply src/xdocs/config_metrics.xml]: Done

File: src/xdocs/config_metrics.xml (hunk 9 above)
Comment (akki111singh) [reply src/xdocs/config_metrics.xml]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java [hunk 15]
```diff
@@ -74,6 +74,23 @@
...
+ * <p>Code Example:</p>
+ * <pre>
+ * public class Test
+ * {
+ *  public static void main(String ... args)
+ *  {
+ *   boolean a = true;
+ *   boolean b = false;
+ *
+ *   boolean c = (a &amp; b) | (b &amp; a); // OK, &amp; and | are counted as 1 which gives
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java]: `// OK, 1(&amp;) + 1(|) + 1(&amp;) = 3`

It is more compact and more descriptive.

Please apply to other examples.
Remember you can wrap expression on few lines to give more space for comments. Only comment ok/violation should be in the same line, explanation could be around if have no space for it, but ideally on the same line.

After update, do not forget to update image and CLI outputs.

Comment (akki111singh) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java]: @romani done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java [hunk 16]
```diff
@@ -82,6 +99,25 @@
...
+ *  {
+ *    boolean a = true;
+ *    boolean b = false;
+ *
+ *    boolean c = (a &amp; b) | (b ^ a) ^ (a | b) | (b ^ a); // OK, since the complexity here is 7
+ *                                                           // (max allowed 7) as &amp;, | and ^
+ *                                                           //  are counted as 1.
+ *
+ *    boolean d = (a | b) ^ (a | b) ^ (a || b) | (b &amp; a) ^ b; // violation, complexity here is 8
+ *                                                                // (max allowed 7) as  |, ^, &amp;
```
[the remark this answers is not in the document]
Comment (akki111singh) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java [hunk 17]
```diff
@@ -90,9 +126,25 @@
  *   &lt;property name="tokens" value="BXOR,LAND,LOR"/&gt;
  * &lt;/module&gt;
  * </pre>
- *
- * @since 3.4
- */
+ * <p>Code Example:</p>
+* <pre>
```
[the remark this answers is not in the document]
Comment (akki111singh) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java (hunk 15 above)
Comment (akki111singh) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/BooleanExpressionComplexityCheck.java]: @romani Can you please review and let me know if any further changes are required.Thanks!
